### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by MattWilliamsDev**
This implements various suggestions/cleanup items from PR comments found at https://developertown.visualstudio.com/Temp%20Check/_git/temp-check/pullrequest/15520?_a=overview

### CODE REVIEW

**Comment by jasonvasquez** [comment]
Github won't let me comment on lines that aren't technically part of the change :)  Could you please prefix the package name with "@developertown/"

File: src/types/Bluetooth.ts [hunk 1]
```diff
@@ -89,3 +44,4 @@ export interface BluetoothConnections {
 
 export const HEALTH_THERMOMETER_UUID = "health_thermometer";
 export const TEMPERATURE_MEASUREMENT_UUID = "temperature_measurement";
```
Comment (abillingsley) [inline src/types/Bluetooth.ts:46]: these seem out of scope of the generic library

File: src/utils/getFloatValue.ts [hunk 2]
```diff
@@ -1,4 +1,9 @@
 // https://gist.github.com/mold/42935cb1bdda7ae9b3ec72e9d0fa8666
```
Comment (abillingsley) [inline src/utils/getFloatValue.ts:1]: are there tests that can be written for these utilities?

File: package.json [hunk 3]
```diff
@@ -65,13 +65,11 @@
     "react-dom": "^16.13.1",
     "react-is": "^16.13.1",
     "react-scripts": "^3.4.1",
+    "react-test-renderer": "^16.13.1",
     "ts-loader": "^7.0.4",
     "tsdx": "^0.13.2",
     "tslib": "^1.12.0",
     "typescript": "^3.9.2",
     "web-bluetooth-mock": "^1.0.2"
-  },
```
[the remark this answers is not in the document]
Comment (MattWilliamsDev) [reply package.json:73]: I've used the `depcheck` utility to find missing/unused deps and the only ones that say they aren't currently used are the jest-related items. I left those in though.

File: src/index.tsx [hunk 4]
```diff
@@ -4,72 +4,62 @@ import {
...
-  }
-
-  // console.log("> " + a.join(" "));
-  return a.join(" ");
-}
-
 export const defaultHookOptions = {
   serviceUuid: HEALTH_THERMOMETER_UUID,
   characteristicUuid: TEMPERATURE_MEASUREMENT_UUID,
+  deviceOptions: { filters: [{ services: [HEALTH_THERMOMETER_UUID] }] },
```
[the remark this answers is not in the document]
Comment (MattWilliamsDev) [reply src/index.tsx:16]: I will remove this default

File: src/utils/getFloatValue.ts (hunk 2 above)
Comment (MattWilliamsDev) [reply src/utils/getFloatValue.ts:1]: This is still very much in flux. @jasonvasquez and I have been discussing solutions for this over the last couple of days. Whatever the solution ends up, there will be tests for it. Should I remove this as the default in the meantime?

**Comment by abillingsley** [comment]
Was chatting with @jasonvasquez in slack and he proposed using a React context at least in part with this (and maybe the speech lib) something like this

```ts
export const BluetoothProvider: React.FC<Props> = ({ children, ...props }) => {
   const bluetooth = useBluetoothNotifications(props)
   return (
    <BluetoothContext.Provider value={bluetooth}>
      {children}
    </BluetoothContext.Provider>
  );
}
export useBluetooth = () => {
  const context = useContext(BluetoothContext);
  if (context === undefined) {
    throw new Error("useBluetooth must be used within a BluetoothProvider");
  }
  return context;
}
```

why? you ask

In simple apps it might be good enough to have just the `useBluetoothNotification` hook but in more complex apps we might have multiple components in the tree that care about bluetooth state

```
<BluetoothProvider>
  <App>
     <ConnectionStatus />
     <TempCheckUI>
         <StartBluetoothButton />
         ....
     </TempCheckUI>
  </App>
</BluetoothProvider>
```

with what we have we could prop drill all the state and callbacks to the appropriate components or with context the components where ever they exist in the tree could simply `useBluetooth` to get the data they care about


This could be shipped as part of this lib or we could make consumers roll their own context.  Figured I'd share what we were discussing here for consideration

**Comment by MattWilliamsDev** [comment]
This is a good idea but perhaps we could add it in a new PR?

**Comment by MattWilliamsDev** [comment]
Though my next thought about the context is that if we are outsourcing notification stream state handling to the implementing component, shouldn't we do the same with context? If that's the case, that would arguably not be part of this lib.

**Comment by abillingsley** [comment]
I think those are slightly different concerns.  Context enables a better access to the bluetooth state when you have a complex component hierarchy (removes the need for prop drilling).  The props passed to context _could_ include the parser function since that is already exposed as an option to the useBluetoothNotification hook.  The difference I see is that as the author of the library we cannot predict the specific bluetooth stream users will require and therefore we need to expose an interface to allow consumers of this library to potentially provide their own implementation for parsing the stream.  However, we can predict that consumers of this library will want to use the same bluetooth state across many components

**Comment by MattWilliamsDev** [comment]
I'm not sure I totally see why that means it should be in this library. An implementor could do all of those things now, I believe, given that this is specific to GATT characteristics. Is there any possible way for the data to come out of it other than as a `(data: DataView, event: Event) => void`?

File: src/BluetoothContext.tsx [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import * as React from "react";
+import { BluetoothConnections, BluetoothNotificationsStatus } from "./types/Bluetooth";
+
+export const BluetoothContext = React.createContext<BluetoothConnections>({
```
Comment (abillingsley) [inline src/BluetoothContext.tsx:4]: See https://kentcdodds.com/blog/how-to-use-react-context-effectively for why a default value should **not** be defined for the context

In my own words, this will mask incorrect usage of the provider and useBluetooth hook.  If we provide no default (undefined) then the useBluetooth hook will throw a nice error if called from outside the BluetoothProvider component.  

This simply becomes
```
export const BluetoothContext = React.createContext<BluetoothConnections | undefined>(undefined)
```

File: src/BluetoothProvider.tsx [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import * as React from "react";
+import { BluetoothNotificationsHookOptions } from "./types/Bluetooth";
+import { BluetoothContext } from "./BluetoothContext";
+import { useBluetoothNotifications } from "./useBluetoothNotifications";
+
+export const BluetoothProvider: React.FC<React.PropsWithChildren<BluetoothNotificationsHookOptions>> = ({
```
Comment (abillingsley) [inline src/BluetoothProvider.tsx:6]: minor but `React.FC` already extends the provided type with `children` so the extra `React.PropsWithChildren` generic type should not be needed

The following should be enough
```
export const BluetoothProvider: React.FC<BluetoothNotificationsHookOptions> = ({
```

File: src/examples/bluetooth-display.tsx [hunk 7]
```diff
@@ -0,0 +1,128 @@
+import * as React from "react";
```
Comment (abillingsley) [inline src/examples/bluetooth-display.tsx:1]: github autocommented on a few things in this file.

File: src/useBluetoothNotifications.tsx [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+  const bluetooth: Bluetooth = useMemo(() => {
+    if (!navigator.bluetooth) {
+      const error = new Error("This device is not capable of using bluetooth");
+      if (onError) {
+        onError(error);
+      } else {
+        throw error;
+      }
+    }
```
Comment (abillingsley) [inline src/useBluetoothNotifications.tsx:53]: we have handleError now.  use that here?

File: src/useBluetoothNotifications.tsx [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+    setDevice(undefined);
+    setStatus(BluetoothNotificationsStatus.STOPPED);
+  }, [server]);
+
+  const startStream = useCallback(async () => {
+    try {
+      setStatus(BluetoothNotificationsStatus.STARTING);
+      let service: BluetoothRemoteGATTService;
+      let characteristic: BluetoothRemoteGATTCharacteristic;
```
Comment (abillingsley) [inline src/useBluetoothNotifications.tsx:99]: these variables are never reassigned so they could be declared as const in the `if (server)` block below

File: src/useBluetoothNotifications.tsx [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        setStatus(BluetoothNotificationsStatus.STOPPING);
+        await characteristic.stopNotifications();
+        reset();
+      }
+    } catch (error) {
+      handleError(error);
+    }
+  }, [characteristic, device, handleError, reset]);
+
+  const _onServerDisconnect = useCallback(
```
Comment (abillingsley) [inline src/useBluetoothNotifications.tsx:140]: nitpicky but this is the method that is prefixed with `_` but it is not the only "private" method in this hook.  based on other conventions used maybe this should be called `handleServerDisconnect`

File: src/BluetoothProvider.tsx (hunk 6 above)
Comment (MattWilliamsDev) [reply src/BluetoothProvider.tsx:6]: That's very helpful to know! Thanks!

File: src/useBluetoothNotifications.tsx (hunk 8 above)
Comment (MattWilliamsDev) [reply src/useBluetoothNotifications.tsx:53]: Feelings on using `invariant` for something like this? I believe doing so allows for extra logging features

File: src/useBluetoothNotifications.tsx (hunk 10 above)
Comment (MattWilliamsDev) [reply src/useBluetoothNotifications.tsx:140]: Good call. "Can't see the forest from the trees" situation on this one I think.

File: src/useBluetoothNotifications.tsx (hunk 8 above)
Comment (abillingsley) [reply src/useBluetoothNotifications.tsx:53]: sure invariant seems fine.  But I do find myself wondering now if we should just require the onError prop? are we setting up consumers of this library for runtime exceptions by making this optional?

File: src/examples/bluetooth-display.tsx (hunk 7 above)
Comment (MattWilliamsDev) [reply src/examples/bluetooth-display.tsx:1]: This file is still in flux until we have a finalized `parser` function. I've intentionally left a couple of things in here for the time being.

File: src/useBluetoothNotifications.tsx (hunk 8 above)
Comment (MattWilliamsDev) [reply src/useBluetoothNotifications.tsx]: If they are using something like a fallback `ErrorBoundary` to catch all `thrown` errors, would having a way to allow the Boundary to catch it instead be desired?
